### ISSUE TITLE
tf.case giving unexpected result in TF 1.0.1

### ISSUE THREAD

**Issue by jwayne** [tensorflow/tensorflow#8776]
### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?

I posted this SO question on 2017-03-10 which was never answered: http://stackoverflow.com/questions/42728235/tensorflow-why-is-tf-case-giving-me-the-wrong-result

### Environment info
Operating System: `Linux 312e492cd9df 4.4.0-66-generic #87-Ubuntu SMP Fri Mar 3 15:29:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`

Installed version of CUDA and cuDNN: none

Installed from: I'm running this on official tensorflow-devel Docker image for 1.0.1 (`gcr.io/tensorflow/tensorflow:1.0.1-devel`)

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)

```
import tensorflow as tf

global_step = tf.Variable(0, dtype=tf.int64)
train_op = tf.assign(global_step, global_step + 1)

learning_rate = tf.Variable(0.1, dtype=tf.float32, name='learning_rate')

# Update the learning_rate tensor conditionally
# When global_step == 2, update to 0.01
# When global_step == 4, update to 0.001
case_tensors = [
    (tf.equal(global_step, 2), tf.constant(0.01, dtype=tf.float32)),
    (tf.equal(global_step, 4), tf.constant(0.001, dtype=tf.float32)),
]
cases = [(pred, lambda: fn_tensor) for pred, fn_tensor in case_tensors]
update = tf.case(cases, default=lambda: learning_rate)
updated_learning_rate = tf.assign(learning_rate, update)

print tf.__version__
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for _ in xrange(6):
        print sess.run([global_step, case_tensors, learning_rate, update, updated_learning_rate])
        sess.run(train_op)
```

### What other attempted solutions have you tried?

None

### Logs or other output that would be helpful
(If logs are large, please upload as attachment or provide link).

The above code prints the following output:
```
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE3 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
1.0.1
[0, [(False, 0.0099999998), (False, 0.001)], 0.1, 0.1, 0.1]
[1, [(False, 0.0099999998), (False, 0.001)], 0.1, 0.1, 0.1]
[2, [(True, 0.0099999998), (False, 0.001)], 0.001, 0.001, 0.001]
[3, [(False, 0.0099999998), (False, 0.001)], 0.001, 0.001, 0.001]
[4, [(False, 0.0099999998), (True, 0.001)], 0.001, 0.001, 0.001]
[5, [(False, 0.0099999998), (False, 0.001)], 0.001, 0.001, 0.001]
```

I expect that the learning rate should get set to `0.0099999998` when the global step reaches 2.  However, even though the predicate for global_step==2 evaluates to True, the learning rate does not get set to `0.0099999998`, but rather gets set to `0.001` instead.

**Comment by aselle** [comment]
```
case_tensors = [
    (tf.equal(global_step, 2), tf.constant(0.01, dtype=tf.float32)),
    (tf.equal(global_step, 4), tf.constant(0.001, dtype=tf.float32)),
]
cases = [(pred, lambda: fn_tensor) for pred, fn_tensor in case_tensors]
```
I believe this is your problem. You are constructing a lambda of already created op handles. You need to give the conditional control over when to actually create the ops... i.e.
```
cases = [
    (tf.equal(global_step, 2), lambda: tf.constant(0.01, dtype=tf.float32)),
    (tf.equal(global_step, 4), labmda: tf.constant(0.001, dtype=tf.float32)),
]
```
You may also need a control dependency between global_step and your case. Honestly, I'd recommend using tf.where for this type of usage anyways.

**Comment by jwayne** [comment]
Great point, thanks. When the lambdas are constructed properly, I get the expected behavior. Please close.

**Comment by jwayne** [comment]
Incidentally, would you be able to explain why defining the lambda with an already-constructed tensor gives a poor result? Or how I might be able to use tf.where to achieve this result of:

if global_step == 2, then set learning_rate = 0.01
if global_step == 4, then set learning_rate = 0.001
else, keep learning_rate the same

**Comment by jwayne** [comment]
Ah - I figured out the tf.where usage. But would still be curious as to the first question.

**Comment by jwayne** [comment]
In particular, if I want it such that if no condition is met, then I want to keep learning_rate the same, then this seems difficult with both tf.where and tf.case (the latter because I cannot let a lambda return an already-constructed tensor)

**Comment by jwayne** [comment]
OK, back to the original bug.  This is what weirds me out.  `test(behavior=1)` gives the unexpected behavior, while `test(behavior=2)` gives the expected behavior:

```
def test(behavior):
    global_step = tf.Variable(0, dtype=tf.int64)
    train_op = tf.assign(global_step, global_step + 1)

    learning_rate = tf.Variable(0.1, dtype=tf.float32, name='learning_rate')

    if behavior == 1:
        case_tensors = [
            (tf.equal(global_step, 2), tf.constant(0.01, dtype=tf.float32)),
            (tf.equal(global_step, 4), tf.constant(0.001, dtype=tf.float32)),
        ]
        cases = [(pred, lambda: fn_tensor) for pred, fn_tensor in case_tensors]
    elif behavior == 2:
        x = tf.constant(0.01, dtype=tf.float32)
        y = tf.constant(0.001, dtype=tf.float32)
        cases = [
            (tf.equal(global_step, 2), lambda: x),
            (tf.equal(global_step, 4), lambda: y),
        ]
    update = tf.case(cases, default=lambda: learning_rate)
    updated_learning_rate = tf.assign(learning_rate, update)

    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())
        for _ in xrange(6):
            print sess.run([global_step, update, updated_learning_rate])
            sess.run(train_op)
```

In particular, the output of `test(1)`:

```
[0, 0.1, 0.1]
[1, 0.1, 0.1]
[2, 0.001, 0.001]
[3, 0.001, 0.001]
[4, 0.001, 0.001]
[5, 0.001, 0.001]
```

The output of `test(2)`:

```
[0, 0.1, 0.1]
[1, 0.1, 0.1]
[2, 0.0099999998, 0.0099999998]
[3, 0.0099999998, 0.0099999998]
[4, 0.001, 0.001]
[5, 0.001, 0.001]
```

**Comment by jwayne** [comment]
Another example where `tf.case` gives unexpected behavior, this time looking at the behavior of `exclusive=False`:

Code:
```
    orig_label = tf.constant(0.046026)
    label_bounds = [0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08]
    preds = [tf.less(orig_label, bound) for bound in label_bounds]
    cases = [(pred, lambda: tf.constant(value))
             for value, pred in enumerate(preds)]
    case = tf.case(cases,
            default=lambda: tf.constant(len(preds)),
            exclusive=False)

    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())
        print sess.run([orig_label, preds, case])
    print list(enumerate(label_bounds))
    print len(preds)
```

Output:
```
[0.046025999, [False, False, False, False, True, True, True, True], 7]
[(0, 0.01), (1, 0.02), (2, 0.03), (3, 0.04), (4, 0.05), (5, 0.06), (6, 0.07), (7, 0.08)]
8
```

(I expect that the first line of the output should read `[0.046025999, [False, False, False, False, True, True, True, True], 4]`, based on https://www.tensorflow.org/api_docs/python/tf/case)

**Comment by aselle** [comment]
The conditionals need to have access to create the ops themselves because they need to construct them and connect them to control dependencies I believe. Thus, you must let the control structure function have control on when the actual ops are created.

**Comment by jwayne** [comment]
Thanks @aselle, that makes sense.  Any word on the unexpected behavior when `exclusive=False` in the post above?

https://github.com/tensorflow/tensorflow/issues/8776#issuecomment-290175897

I just confirmed that this behavior also occurs in the nightly build.  Will get confirmation from colleague on whether he also thinks the behavior is unexpected @knub

**Comment by knub** [comment]
I would also expect `case` to evaluate to `4`.

**Comment by aselle** [comment]
@yuanbyu, @jwayne's example above seems to be a inconsistency in the documentation and exact behavior with `exclusive=False`.

**Comment by jwayne** [comment]
@itsmeolivia The bug described in [this comment](https://github.com/tensorflow/tensorflow/issues/8776#issuecomment-290175897), confirmed by @aselle and @knub, still exists in the latest tensorflow version.

I can work on a patch - can I open a new issue with a cleaner history, referencing this issue, and continue that way?

**Comment by itsmeolivia** [comment]
yes, go for it.  I initially closed the issue because no one had commented on over 2 months but if you're still interested in finding a solution then have at it.

**Comment by kopekC** [comment]
This is not a TF issue.

Closures are defined over names and not over values (https://stackoverflow.com/a/13355291). In this case, all the lambdas ended up with the value the variable had at the end of the loop. A simple solution is to have a lambda generate the lambdas you are trying to iterate over and pass the iterated value as the first and only argument. The closure is now defined over a constant, or the result of the evaluation of the first lambda (which only depends on the argument, which is itself copied).

So the code that yields the results you are looking for looks like:

    import tensorflow as tf

    orig_label = tf.constant(0.046026)
    label_bounds = [0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08]
    preds = [tf.less(orig_label, bound) for bound in label_bounds]
    # Pair each predicate with a function returning the index of the predicate.
    z = lambda x: lambda: tf.constant(x)
    pred_fn_pairs = [(pred, z(i))
                     for i, pred in enumerate(preds)]
    # If no predicate evaluates to true, default to returning the index after
    # the index of the last predicate.
    default = lambda: tf.constant(len(pred_fn_pairs))
    case = tf.case(pred_fn_pairs, default=default, exclusive=False)

    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())
        print(sess.run([preds, case]))
    print(list(enumerate(label_bounds)))
    print(len(preds))

**Comment by jwayne** [comment]
@knub and I agree with @kopekC. The issue described is not a TF issue, but rather an issue with how the problem was constructed, and this thread should remain closed.
